Camera App changes to stitch Stream sources(GStreamer based) to Transports

pidarped opened the pull request:
* Define Camera HAL interface methods(start/stop streams and initial getter/setters of camera h/w attributes)
* Define CamerAVStreamMgmt Delegate for handling Stream alocation/de-allocation/start/stop using HAL Impl.
* Gstreamer based A/V stream creation to designated network sink.
* Define Network stream sources(Video/Audio) that listen on the socket and distribute streams to a MediaController.
* Define a Transport interface for sending Audio/Video.
* Define a MediaController that Stream transports can subscribe to/unsubscribe from.
* WebRTC and PushAV implement the Transport interface.
* When stream is running, media-controller fetches streams and distributes to registered transports.

#### Testing
* Allocation commands from chip-tool
* CaptureSnapshot command using allocated Snapshot streamID

pidarped committed:
Initial delegate for camera-app.

pidarped committed:
Add camera-hal interface.

pidarped committed:
Camera App changes

- Define Camera HAL interface inside camera-device-interface.h.
- Move stream handling logic from AVStream Delegate to inside CameraDevice.
- Add initial Gstreamer support inside CameraDevice for stream
  start/stop.

pidarped committed:
Add camera-av-stream-manager.h header

pidarped committed:
Remove camera-hal.h since the interface methods are integrated into camera-device-interface.h.

pidarped committed:
Fixes to gstreamer pipelines and Camera AVStreamManager Init

pidarped committed:
Add ResourceExhausted error return for StreamAllocate functions.

pidarped committed:
Restyle fixes

pidarped committed:
Refactor snapshot pipeline.

-Add separate snapshot pipeline creation helper function.
-Have stream started at Init.
-Capture snapshot reads latest copy of image file.

pidarped committed:
Remove glib-2.0 from dependencies.

--Also ensure the resolution values of the correct snapshot stream id is
returned in the CaptureSnapshot response.

pidarped committed:
Use pkg_config to configure BUILD.gn for gstreamer instead of manually including headers and libs.

pidarped committed:
Add initial media controller and transport interface headers.

pidarped committed:
Add NetworkStreamSource and MediaController logic.

NetworkStreamSource is used to source each stream from a network source
and distribute to the media controller.

MediaController multiplexes the stream to registered transports that are
ready to receive it.

Separate out stream source distribution to transports based on type.

pidarped committed:
Add more checks to stream allocation functions.

pidarped committed:
Move stream allocation/de-allocation function logic from HAL interface
to CameraAVStreamManager.

Streams are pre-allocated by CameraDevice(implementer of HAL) and
allocation/de-allocations are managed by camera-av-stream-mgmt.
Camera-av-stream-mgmt uses HAL interface to start/stop streams.

pidarped committed:
Address review comments.

- Loop on select on the socket with small sleeps on timeout.
- Make the CameraDevice not be a singleton and use a setter on
  CameraAVStreamManager to set it, so that the HAL functions can be
  accessed via the member CameraDevice object instead of a singleton
  GetIntance() reference.
- Move the Stream struct definitions into the CameraHAL with a
  context pointer to hold platform-specific stream context(e.g.,
  GstElement)
- Make the transportActive bool atomic.
- Define utility method for streams to compare allocation parameters.

pidarped commented on examples/camera-app/linux/src/network-stream-source/network-stream-source.cpp:
```diff
@@ -0,0 +1,105 @@
+    if (bind(serverSocket, reinterpret_cast<sockaddr *>(&serverAddress), sizeof(serverAddress)) == -1)
+    {
+        close(serverSocket);
+        ChipLogError(NotSpecified, "Failed to bind socket.");
```
Yes, added errno to all error logs.

pidarped commented on examples/camera-app/linux/include/network-stream-source/network-stream-source.h:
```diff
@@ -0,0 +1,51 @@
+    uint16_t mSrcPort;
+    uint16_t mStreamId;
+    StreamType mStreamType;
+    bool mStreamSourceActive = false;
```
Made it std::atomic.

pidarped commented on examples/camera-app/linux/src/network-stream-source/network-stream-source.cpp:
```diff
@@ -0,0 +1,105 @@
+    while (mStreamSourceActive)
+    {
+        ssize_t bytesReceived =
+            recvfrom(serverSocket, buffer, sizeof(buffer), 0, reinterpret_cast<sockaddr *>(&clientAddress), &clientAddressLength);
```
Changed to a select based socket monitoring loop with sleeps in between. Initially, I had thought that the blocking call was probably ok in a separate thread. Thanks for pointing to the issue with Stop().

pidarped commented on examples/camera-app/linux/src/clusters/camera-avstream-mgmt/camera-av-stream-manager.cpp:
```diff
@@ -0,0 +1,299 @@
+        {
+            foundAvailableStream = true;
+
+            if (stream.codec == allocateArgs.videoCodec && stream.frameRate >= allocateArgs.minFrameRate &&
```
Yes. This was a mental Todo to have the checks handled by a utility method. Have refactored them.

pidarped commented on examples/camera-app/linux/src/clusters/camera-avstream-mgmt/camera-av-stream-manager.cpp:
```diff
@@ -0,0 +1,299 @@
+        if (stream.id == streamID && stream.isAllocated)
+        {
+            // Stop the video stream
+            CameraDevice::GetInstance().StopVideoStream(stream.id);
```
Created a helper method in CameraAVStreamManager to set the CamerDeviceHAL interface to use to access the HAL functions. Needed to do some refactoring to allow that, e.g., moving the available stream definitions inside the HAL(which makes sense).

pidarped commented on examples/camera-app/linux/include/camera-device.h:
```diff
@@ -17,26 +17,147 @@
+    {
+        static CameraDevice sCameraDevice;
+        return sCameraDevice;
+    }
```
Having set a pointer to the CameraDevice object in the CameraAVStreamManager object, the former is no longer defined as a singleton.

pidarped committed:
Update linux standalone yaml to latest Docker version.

Also update Log module to `Camera` in the camera-app.

yufengwangca commented on examples/camera-app/camera-common/src/camera-app.cpp:
```diff
@@ -21,16 +21,60 @@ using namespace chip;
+    optionalAttrs.Set(chip::app::Clusters::CameraAvStreamManagement::OptionalAttribute::kNightVisionIllum);
+    uint32_t maxConcurrentVideoEncoders  = mCameraDevice->GetCameraHALInterface().GetMaxConcurrentVideoEncoders();
+    uint32_t maxEncodedPixelRate         = mCameraDevice->GetCameraHALInterface().GetMaxEncodedPixelRate();
+    VideoSensorParamsStruct sensorParams = mCameraDevice->GetCameraHALInterface().GetVideoSensorParams();
```
Those initialization are not platform independent, should be moved to linux

pidarped replied:
These are accessing the abstract Camera HAL Interface functions. The implementation of those functions would be platform-specific.

yufengwangca replied:
Acked

yufengwangca commented on examples/camera-app/linux/include/clusters/camera-avstream-mgmt/camera-av-stream-manager.h:
```diff
@@ -0,0 +1,86 @@
+    Protocols::InteractionModel::Status VideoStreamModify(const uint16_t streamID, const chip::Optional<bool> waterMarkEnabled,
+                                                          const chip::Optional<bool> osdEnabled);
+
+    Protocols::InteractionModel::Status VideoStreamDeallocate(const uint16_t streamID);
```
Add overload to those overloaded functions to force signature check

pidarped replied:
Did you mean override? Added that in the delegate.

pidarped committed:
Remove snapshot file when stream is deallocated and other fixes.

-Add `override` keyword to implemented delegate methods.
-Directly access stream vector from HAL in loop.
-Remove unnecessary gst header inclusion.

pidarped commented on examples/camera-app/linux/BUILD.gn:
```diff
@@ -29,18 +31,35 @@ config("config") {
+  configs += [ ":gstreamer" ]
+
   sources = [
+    "${chip_root}/examples/camera-app/linux/include/media-controller/media-controller.h",
```
The current implementation uses constructs that make it more suitable to linux based platforms and so I chose to keep it under linux and not outright put it in common. We can revisit this later to see if some or all of this portion can be moved to common.

yufengwangca replied:
Ok, we can refactor this later until all basic pieces are ready

pidarped commented on examples/camera-app/linux/include/media-controller/media-controller.h:
```diff
@@ -0,0 +1,48 @@
+    void DistributeAudio(const char * data, size_t size, uint16_t audioStreamID);
+
+private:
+    std::vector<Connection> connections;
```
Yes, the transports need to be unregistered from the media-controller before being destroyed. Have added comments in the header to clarify. Keeping them as pointers for now.
